References

aapeliv opened the pull request:
This is a complete rewrite of the references system to incorporate non-friend references after host requests.

aapeliv committed:
Separate references into its own service

aapeliv committed:
Fix up

aapeliv committed:
Fix tests

aapeliv committed:
Store dates as dates, not strings

aapeliv committed:
Clean up

aapeliv committed:
Move parse_date to utils, not db for consistency

aapeliv committed:
Fix name clash

aapeliv committed:
Fix up references table

aapeliv committed:
Fix migrations

aapeliv committed:
Work on references API

aapeliv committed:
Start on tests

aapeliv committed:
Test pagination listing

aapeliv committed:
Fix up reference timezone handling

aapeliv committed:
Clean up

aapeliv committed:
Finish tests

aapeliv committed:
Condense pending reference queries into one with UNION ALL

aapeliv committed:
Clean up

aapeliv committed:
Send emails about references

kalvdans commented on app/backend/src/couchers/errors.py:
```diff
@@ -61,6 +62,7 @@
 MISSING_PAGE_PARENT = "Missing page parent."
 MISSING_PAGE_TITLE = "Missing page title."
 MISSING_PASSWORD = "The user has a password but the request did not provide one."
+NEED_TO_SPECIFY_AT_LEAT_ONE_USER_ID = "You need to specify at least one user id."
```
spelling

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -655,23 +625,107 @@ class HostRequest(Base):
-    # dates as "YYYY-MM-DD", in the timezone of the host
-    from_date = Column(String, nullable=False)
-    to_date = Column(String, nullable=False)
+    from_date = Column(Date, nullable=False)
```
document timezone

kalvdans commented on app/backend/src/couchers/tasks.py:
```diff
@@ -102,6 +102,40 @@ def send_friend_request_email(friend_relationship):
 
+def send_host_reference_email(reference, both_written):
+    """
+    both_written iif both the surfer and hoster wrote a reference
```
iff*

kalvdans commented on app/backend/src/couchers/utils.py:
```diff
@@ -19,6 +20,20 @@ def Timestamp_from_datetime(dt: datetime):
     return pb_ts
 
 
+def parse_date(d: date):
```
d: str

aapeliv replied:
Good catch, thanks!

kalvdans commented on app/pb/references.proto:
```diff
@@ -0,0 +1,104 @@
+    // Raises FAILED_PRECONDITION if you try to write a reference for someone you already wrote a friend reference for.
+  }
+
+  rpc WriteHostRequestReference(WriteHostRequestReferenceReq) returns (Reference) {
```
s/host request completed/surfed/

aapeliv replied:
I clarified a bit. This same function is used both when surfing or hosting

kalvdans reviewed (commented):
some comments from mobile :)

aapeliv committed:
Clarify things and fix typos

aapeliv committed:
Clarify tzs

aapeliv commented:
Closes #413

aapeliv commented:
Regarding timezones: the idea is that each table that contains local times also contains the tzdata identifier for which timezone those dates/times are in. Mostly that timezone identifier should be copied from the timezone identifier from the creating user, whose timezone identifier is set automatically from their location — see #795 (or overwritten manually, maybe?). This is more important for things like events that handle times and not just dates: but I thought it'd be nice have the time you can write references be until midnight local time (since it'll be visible in the UI) instead of like UTC. And also it's good that people can only start writing references after they've hosted, so midnight of the to_date. (Though now thinking about it, there might be an off-by-one error here).

aapeliv committed:
Add some more dummy references and fix up dummy data

aapeliv committed:
Add one more test to test second host ref

lucaslcode commented on app/pb/references.proto:
```diff
@@ -0,0 +1,104 @@
+  string next_page_token = 2;
+}
+
+message WriteFriendReferenceReq {
+  int64 to_user_id = 1;
+  string text = 2;  // plain text
+  bool was_safe = 3;
+  int32 rating = 4;
```
Based on the mockups, I think this should have `bool was_comfortable` (maybe appropriate is a better word than comfortable, though?) and rating should be a float from 0-1. I am going to ask for clarification in the slack.

lucaslcode replied:
@aapeliv

aapeliv replied:
Done.

darrenvong committed:
feat: update reference call + add pagination support

darrenvong committed:
feat: refactor references view stuff into more components

darrenvong committed:
feat: rename hasPages to hasAtLeastOnePage

darrenvong committed:
feat: update/add/fix tests and storybook for references

darrenvong committed:
feat: add tests for hasAtLeastOnePage utility

darrenvong committed:
feat: increase pageSize to maximum supported

darrenvong committed:
feat: lower number of references fetched per page

darrenvong committed:
feat: remove all references list to simplify code

darrenvong commented:
@aapeliv Just a heads up I merged in the frontend portion for working with this new refactored references API

darrenvong commented:
Got lazy and fixed the frontend conflicts directly in the web editor, hopefully I didn't f*** anything up...

aapeliv committed:
Update was_safe and rating to [0,1]

lucaslcode commented on app/pb/references.proto:
```diff
@@ -72,15 +72,17 @@ message ListReferencesRes {
   string text = 2;  // plain text
-  bool was_safe = 3;
-  int32 rating = 4;
+  bool was_appropriate = 3;
```
Sorry I meant keep `was_safe` and add `was_appropriate`...

aapeliv committed:
Fix rating

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -655,23 +625,113 @@ class HostRequest(Base):
+    rating = Column(Float, nullable=False)
+    was_appropriate = Column(Boolean, nullable=False)
+
+    visible_from = Column(DateTime(timezone=True), nullable=False)
```
Can be calculated from `time`, no need to store in db.

aapeliv replied:
The reason I've added it is that it's really quite annoying to compute. Logic is approximately this:

1. If it's a friend reference, it's immediately visible
2. Otherwise, it's visible if the other reference from the person you surfed/hosted has written the reference OR if the reference is over 2 weeks old

Storing this means it's very easy to get all visible references, avoiding a few hairy joins in some places. But I can also remove this if you feel it's important.

Also, currently it's ready for timezone support, which would need to included in that join!

kalvdans replied:
It would be more annoying to remember to recalculate the whole `visible_from` column on every friend/unfriend/reference written/timezone change. Cache invalidation is the second hardest problem in computer science :)

aapeliv replied:
I'll change it.

But this is set once, except for a hosting/surifng reference if the other person writes a reference within 2 weeks. Doesn't get updated if you friend/unfriend/tz change,etc.

kalvdans replied:
Well if visible_from should be based on an old snapshot of the database then we can't calculate it on demand, then we need a column for it. Just have to make up our mind :D

aapeliv replied:
Done, turned into a subquery: I agree, it's cleaner this way! Thanks :)

kalvdans reviewed (commented):
Fixing timezone later will be a pain after introducing things like `today()`... Should be addressed in this PR in my opinion.

aapeliv commented:
> Fixing timezone later will be a pain after introducing things like `today()`...

`today()` is not used much in the actual code, it's mostly in tests, so I think it's fine.

> Should be addressed in this PR in my opinion.

I'd prefer to fix all of that up in another PR where we introduce tzs for users, which then go into host requests, and from there into here. The main stuff left in this is the `least_current_date` which needs to be completely removed. I guess we need a `today_in_tz(tz=...)` function instead.

Most of this is otherwise tz aware already

kalvdans commented:
Sounds like you have a plan how to tackle the timezone later, so I take your word for it :)

aapeliv committed:
Clean up

aapeliv committed:
Clean up tz handling

aapeliv committed:
Fix up tz handling

aapeliv committed:
Fix typo

aapeliv committed:
Get rid of visible_from

aapeliv commented:
@kalvdans: I removed `visible_from`, and agree now, it's cleaner. I also cleaned up a bunch of the timezone stuff, but will leave it to another PR to implement non-UTC stuff.

Ready for re-review now!

kalvdans reviewed (approved):
Backend looks fine now!
